### ISSUE TITLE
[RELEASE-Proposal] PRs to merge before we cut a new release for halo2

### ISSUE THREAD

**Issue by CPerezz** [privacy-scaling-explorations/halo2#82]
Considering the amount of changes we currently have on `HEAD` + the upcomming PRs we should consider to cut a new release and integrate into `zkevm-circuits`.

This is the list of PRs I would merge before we cut the release:
- [ ] #77 (Contains upstream rebase too. So comes with quite some changes).
- [ ] #76 We can basically get rid of the blinders.
- [ ] #81 Fixes an expression evaluation issue.
- [ ] #78 
- [ ] #79 (If we can have gpu in our CI. Otherwise, this one can wait) cc: @AronisAt79 @ntampakas @barryWhiteHat 

cc: @kilic @han0110

**Comment by CPerezz** [comment]
@kilic I'm not sure I follow. All these PRs were merged into `main`. So I thought we were planning to keep merging into `main` the ones I listed above. 

But it seems that you have a different plan. I guess that the Commitment Scheme abstraction is too hard to merge into `main`?

**Comment by kilic** [comment]
@CPerezz even without the abstraction stuff diff for main  __70 commits ahead 198 commits behind__ against upstream. So this `abstraction` effort is like base for our usual manual rebase. I know it's horrible to manually rebase each time but it looked like most comfortable way to avoid huge rebase conflicts while synching with upstream.

**Comment by CPerezz** [comment]
@kilic no, it's fine. Sometimes there's no better way sadly..

Let's see if @brechtpd @NoCtrlZ @lispc @ntampakas @AronisAt79  and @han0110 can take some time to do this

**Comment by AronisAt79** [comment]
@CPerezz not sure if the above PR-lists contain everything that is associated to halo2 benchmarking via CI. But afaiu, the idea here is to port the complete CI (benchmarking) functionality of our halo2 main to abstraction branch. correct?

**Comment by Brechtpd** [comment]
> @Brechtpd what would you think moving some of polynomial and evaluation related improvements to the upstream?

Sure. With upstream you mean [`zcash/halo2`](https://github.com/zcash/halo2) and not the `abstraction` branch right?

**Comment by kilic** [comment]
> @CPerezz not sure if the above PR-lists contain everything that is associated to halo2 benchmarking via CI. But afaiu, the idea here is to port the complete CI (benchmarking) functionality of our halo2 main to abstraction branch. correct?

@AronisAt79 Yes and I think whole benchmarking work can be a single PR if you would also agree.

> Sure. With upstream you mean [zcash/halo2](https://github.com/zcash/halo2) and not the abstraction branch right?

@Brechtpd Yes I mean zcash/halo2 if you also think it make sense and it's not so much zkevm focused. However we can expect that PR in the upstream might take a while to get merged. So depending to the response it gets we might want to have it on `abstraction` first before it is accepted

**Comment by CPerezz** [comment]
@kilic  https://github.com/zcash/halo2/pull/593 is essential for us to be able to forge keccak and other RLC-dependant circuits.

**Comment by han0110** [comment]
For challenge API I think I have addressed all addressable comments, and left some I don't have clear idea how to deal with. Not sure when it would be reviewed again, but if we want to use it right now I can port it to our fork, it should be really easy after #77 is merged because they are on the same base then.

**Comment by str4d** [comment]
> Not sure when it would be reviewed again

Reviewing it now 🙂

**Comment by kilic** [comment]
How bad we want to include [last 7 PRs](https://github.com/zcash/halo2/pulls?q=is%3Apr+is%3Aclosed) from upstream?. I've tried to rebase but meet with some pesky conflicts. Should we park where `abstraction` is based or we want to fetch them too for this release?

**Comment by CPerezz** [comment]
https://github.com/zcash/halo2/pull/608 Might be the only one. But since we have a bunch of parallelization and GPU opts incoming, I think we can skip all 7 for this release.

**Comment by therealyingtong** [comment]
Do we want `Circuit::Value` (https://github.com/zcash/halo2/pull/598) as well? This will mean updating the types from `Option<V>` to `Value<V>` in every circuit, but I don't think it would introduce many rebase conflicts.

**Comment by CPerezz** [comment]
> Do we want `Circuit::Value` ([zcash#598](https://github.com/zcash/halo2/pull/598)) as well? This will mean updating the types from `Option<V>` to `Value<V>` in every circuit, but I don't think it would introduce many rebase conflicts.

That's my bad. I thought I listed it.. Indeed, I want that TBH. I filed the original issue for it in halo2 upstream.
Will add it too on the top.

**Comment by han0110** [comment]
I have picked all necessary changes and squash them into branch [`feature/abstraction-squashed`](https://github.com/privacy-scaling-explorations/halo2/tree/feature/abstraction-squashed). I didn't force update [`abstraction`](https://github.com/privacy-scaling-explorations/halo2/tree/abstraction) branch because `halo2wrong` depends on it and #29 add another argument `name` to API `ConstraintSystem::lookup` as a breaking change.

I think a better way to process this release is to open another PR and leave [`abstraction`](https://github.com/privacy-scaling-explorations/halo2/tree/abstraction) untouched, then after the release tag is ready, we can start to update everything in `halo2wrong` and `zkevm-circuits`. Will open a PR soon if this makes sense.

**Comment by CPerezz** [comment]
Closed via #93
